### PR TITLE
chart-data flow

### PR DESCRIPTION
**Pull request by Kachyr**
Method for data flow of chart

### CODE REVIEW

File: src/app/repository/repository.service.ts [hunk 1]
```diff
@@ -12,6 +13,28 @@ import TEAMS from '../mocks/teams/teams.json';
   providedIn: 'root',
 })
 export class DataRepository {
+  /**
+   * Method will return array with data for char,
```
Comment (apastuhov) [inline src/app/repository/repository.service.ts:17]: typo: chart

File: src/app/repository/repository.service.ts [hunk 2]
```diff
@@ -12,6 +13,28 @@ import TEAMS from '../mocks/teams/teams.json';
...
+  getChartData(lastRequestDateTime: number): Observable<ChartData[]> {
+    const dataArr: ChartData[] = [];
+    const now = Date.now();
+
+    const timeDifference = lastRequestDateTime
+      ? Math.round((now - lastRequestDateTime) / 1000)
+      : 5;
+    //generate objects for chart
+    for (let i = 0; i < timeDifference; i++) {
+      dataArr.push(this.generateChartObject(i));
```
Comment (apastuhov) [inline src/app/repository/repository.service.ts:31]: pass actual time as a property, not `i`

File: src/app/repository/repository.service.ts [hunk 3]
```diff
@@ -12,6 +13,28 @@ import TEAMS from '../mocks/teams/teams.json';
...
+   * Method will return array with data for char,
+   * number of elements depends depends on difference last request time and current moment
+   * @param lastRequestDateTime time of last request in millisecond format
+   * @returns Observable with array of chart data objects
+   */
+  getChartData(lastRequestDateTime: number): Observable<ChartData[]> {
+    const dataArr: ChartData[] = [];
+    const now = Date.now();
+
+    const timeDifference = lastRequestDateTime
```
Comment (apastuhov) [inline src/app/repository/repository.service.ts:26]: timeDifference - is it in seconds / ms or what? and why we have ternary operator there?

File: src/app/chart/chart.component.ts [hunk 4]
```diff
@@ -1,8 +1,34 @@
...
+    left: number;
+  };
+  lastReq!: number;
+  constructor(
+    private repo: DataRepository,
+    private timerService: TimerService,
+  ) {}
+
+  ngOnInit(): void {
+    // TODO: implement visualization of data
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:23]: remove todo

File: src/app/chart/chart.component.ts [hunk 5]
```diff
@@ -1,8 +1,34 @@
...
+    this.timerService.runningTimer.subscribe((timerData) => {
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:24]: do not forget to unsubscribe

File: src/app/chart/chart.component.ts [hunk 6]
```diff
@@ -1,8 +1,34 @@
...
   templateUrl: './chart.component.html',
   styleUrls: ['./chart.component.css'],
 })
-export class ChartComponent {}
+export class ChartComponent implements OnInit {
+  timerData?: {
+    passed: number;
+    left: number;
+  };
+  lastReq!: number;
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:16]: what is `lastReq`?

File: src/app/chart/chart.component.ts [hunk 7]
```diff
@@ -1,8 +1,34 @@
...
+  constructor(
+    private repo: DataRepository,
+    private timerService: TimerService,
+  ) {}
+
+  ngOnInit(): void {
+    // TODO: implement visualization of data
+    this.timerService.runningTimer.subscribe((timerData) => {
+      if (timerData.passed % 5 === 0 && timerData.passed > 0) {
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:25]: add check to ensure that timer didn't ended (check the `left`) property. also use rxjs operator to do what you want

File: src/app/chart/chart.component.ts [hunk 8]
```diff
@@ -1,8 +1,34 @@
...
+        this.repo.getChartData(this.lastReq).subscribe((data) => {
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:26]: never do subscribe in subscribe.. use `switchMap` or something like it

File: src/app/chart/chart.component.ts [hunk 9]
```diff
@@ -1,8 +1,34 @@
...
+          console.log(data);
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:27]: display data in HTML via comma

File: src/app/chart/chart.component.ts [hunk 10]
```diff
@@ -1,8 +1,34 @@
...
+        });
+        this.lastReq = Date.now();
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:29]: do it before sending the request

**Review by apastuhov — CHANGES_REQUESTED** [review]

File: src/app/repository/repository.service.ts (hunk 1 above)
Comment (Kachyr) [reply src/app/repository/repository.service.ts:17]: done

File: src/app/chart/chart.component.ts (hunk 6 above)
Comment (Kachyr) [reply src/app/chart/chart.component.ts:16]: that is time of the last request

File: src/app/chart/chart.component.ts (hunk 7 above)
Comment (Kachyr) [reply src/app/chart/chart.component.ts]: done

File: src/app/chart/chart.component.ts (hunk 8 above)
Comment (Kachyr) [reply src/app/chart/chart.component.ts]: done

File: src/app/chart/chart.component.ts (hunk 10 above)
Comment (Kachyr) [reply src/app/chart/chart.component.ts]: done

File: src/app/chart/chart.component.ts [hunk 11]
```diff
@@ -1,8 +1,46 @@
...
 @Component({
   selector: 'app-chart',
   templateUrl: './chart.component.html',
   styleUrls: ['./chart.component.css'],
 })
-export class ChartComponent {}
+export class ChartComponent implements OnInit, OnDestroy {
+  timerSub!: Subscription;
+  lastRequestTime!: number;
+  stepOfChart = 5;
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:16]: private readonly + jsdoc

File: src/app/chart/chart.component.ts [hunk 12]
```diff
@@ -1,8 +1,46 @@
...
+          ({ left, passed }) =>
+            passed % this.stepOfChart === 0 && passed >= 0 && left >= 0,
+        ),
+        switchMap((timerIsRunning) =>
+          timerIsRunning ? this.repo.getChartData(this.lastRequestTime) : EMPTY,
+        ),
+      )
+      .subscribe((data) => {
+        this.lastRequestTime = Date.now();
+        this.data = data;
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:39]: this.data.push(...data);

File: src/app/chart/chart.component.ts [hunk 13]
```diff
@@ -11,10 +11,14 @@ import { GameChartData } from '../shared/models/chart.model';
...
-  stepOfChart = 5;
-  data!: any;
+  /**
+   * Value that determines how often request a new chart data
+   * @type {number}
+   */
+  private readonly stepOfChart = 5;
+  private timerSub!: Subscription;
+  private lastRequestTime!: number;
+  private data!: GameChartData[];
```
Comment (apastuhov) [inline src/app/chart/chart.component.ts:21]: if var is private and is used in template - it will not work on prod build.

**Review by apastuhov — APPROVED** [review]
LGTM, thanks

**Pull request merged**